Engine/Component/ComponentTrailRenderer.cpp

```diff
@@ -0,0 +1,179 @@
+#include "ComponentTrailRenderer.h"
+
+#include "Main/Application.h"
+
+#include "Module/ModuleEffects.h"
+#include "Module/ModuleProgram.h"
+#include "Module/ModuleResourceManager.h"
+#include "Module/ModuleRender.h"
+#include "Module/ModuleTime.h"
+
+#include "ResourceManagement/ResourcesDB/CoreResources.h"
+
+namespace { const float MAX_TRAIL_VERTICES = 20; } //arbitrary number 
+
+ComponentTrailRenderer::ComponentTrailRenderer() : Component(nullptr, ComponentType::TRAILRENDERER)
+{
+	InitData();
+}
+
+ComponentTrailRenderer::ComponentTrailRenderer(GameObject * owner) : Component(owner, ComponentType::TRAILRENDERER)
+{
+	InitData();
+}
+
+ComponentTrailRenderer::~ComponentTrailRenderer()
+{
+	if (trail_vbo != 0)
+	{
+		glUnmapBuffer(GL_ARRAY_BUFFER);
+		glDeleteBuffers(1, &trail_vbo);
+		glDeleteBuffers(1, &trail_vao);
+	}
+	App->effects->RemoveComponentTrailRenderer(this);
+}
+
+void ComponentTrailRenderer::InitData()
+{
+	ChangeTexture(static_cast<uint32_t>(CoreResource::BILLBOARD_DEFAULT_TEXTURE));
+
+	glGenVertexArrays(1, &trail_vao);
+	glGenBuffers(1, &trail_vbo);
+
+	glBindVertexArray(trail_vao);
+
+	glBindBuffer(GL_ARRAY_BUFFER, trail_vbo);
+	glBufferData(GL_ARRAY_BUFFER, sizeof(float) * MAX_TRAIL_VERTICES * 4 * 5, nullptr, GL_DYNAMIC_DRAW); 
+	
+	glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 3 * sizeof(float), (void*)0);
+	glEnableVertexAttribArray(0);
+
+	glBindBuffer(GL_ARRAY_BUFFER, 0);
+	glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
+	glBindVertexArray(0);
+}
+
+//void ComponentTrailRenderer::SwitchFrame()
+//{
+//	time_since_start += App->time->delta_time;
+//	APP_LOG_INFO("%.1f", time_since_start);
+//
+//	if (time_since_start * sheet_speed >= 1)
+//	{
+//		current_sprite_x += 1;
+//
+//		if ((int)current_sprite_x >= x_tiles) {
+//			current_sprite_y--;
+//			current_sprite_x = 0;
+//		}
+//
+//		if ((int)current_sprite_y <= 0) {
+//			current_sprite_y = y_tiles;
+//		}
+//		time_since_start = 0.f;
+//	}
+//
+//
+//}
+
+void ComponentTrailRenderer::Render(std::vector<float>& to_render)
+{
+	if (active)
+	{
+		/*for (int i = 0; i < to_render.size(); i += 3)
+	{
+		float3 point = {to_render[i], to_render[i + 1], to_render[i + 2]};
+		App->debug_draw->RenderPoint(point, 20.0f);
+	}*/
+		GLuint shader_program = App->program->GetShaderProgramId("Trail");
+
+		glUseProgram(shader_program);
```

Comment:
With uber shader you need to substitute this 2 lines with `GLuint shader_program = App->program->UseProgram("Trail");`

Reply:
I'm not aware of the uber shader. The Particles and Billboard still use the GetShaderProgramId()  method. Has uber shader been implemented yesterday in develop?